core/src/main/java/tech/tablesaw/io/ReadOptions.java

```diff
@@ -327,8 +357,112 @@ public Builder minimizeColumnSizes() {
+     */
+    public Builder columnTypeByNameFunction(
+        Function<String, Optional<ColumnType>> columnTypeFunction) {
+      this.columnTypeFunction = columnTypeFunction;
+      return this;
+    }
+
+    /**
+     * Provide a function that determines ColumnType for all column names. To provide only for some
+     * use {@link #columnTypeByNameFunction(Function)}
+     *
+     * <p>Providing that function prevents running autodetect column type logic
+     */
+    public Builder completeColumnTypeByNameFunction(
+        Function<String, ColumnType> columnTypeFunction) {
+      this.completeColumnTypeFunction = columnTypeFunction;
+      return this;
+    }
+
+    public Builder columnTypes(Map<String, ColumnType> columnTypeByName) {
+      if (columnTypeByName != null) this.columnTypeMap = columnTypeByName;
+
+      return this;
+    }
+
     public ReadOptions build() {
       return new ReadOptions(this);
     }
   }
+
+  /**
+   * Allow to customize read column types. It can work in three ways:
+   *
+   * <ul>
+   *   <li>If no information is provided column types are autodetected
+   *   <li>A complete list of columns can be provided using {@link
+   *       ReadOptions.Builder#columnTypes(ColumnType[])} or {@link
+   *       ReadOptions.Builder#completeColumnTypeFunction} and they are used preventing autodetect
+   *   <li>Provide values for some column names using {@link
+   *       ReadOptions.Builder#columnType(String,ColumnType)} or {@link
+   *       ReadOptions.Builder#columnTypeByNameFunction(Function)} (String, ColumnType)}. In this
+   *       case provided columnTypes are used and the others are autodetected
+   * </ul>
+   */
+  public static class ColumnTypeReadOptions {
+    final ColumnType[] columnTypesByIdx;
+    final Map<String, ColumnType> columnTypesByNameMap;
+    final Function<String, Optional<ColumnType>> columnTypesByNameFunction;
+    final Function<String, ColumnType> completeColumnTypesByNameFunction;
+
+    public static ColumnTypeReadOptions of(ColumnType[] allColumnTypes) {
+      return new ColumnTypeReadOptions(allColumnTypes, null, null, null);
+    }
+
+    ColumnTypeReadOptions(
+        ColumnType[] columnTypesByIdx,
+        Map<String, ColumnType> columnTypesByNameMap,
+        Function<String, ColumnType> completeColumnTypesByNameFunction,
+        Function<String, Optional<ColumnType>> columnTypesByNameFunction) {
+      this.columnTypesByIdx = columnTypesByIdx;
+      this.columnTypesByNameMap = columnTypesByNameMap;
+      this.columnTypesByNameFunction = columnTypesByNameFunction;
+      this.completeColumnTypesByNameFunction = completeColumnTypesByNameFunction;
+    }
+
+    public Optional<ColumnType> columnType(int columnNumber, String columnName) {
+      Optional<ColumnType> columnType = Optional.empty();
+      if (columnTypesByIdx != null && columnNumber < columnTypesByIdx.length)
+        columnType = Optional.ofNullable(columnTypesByIdx[columnNumber]);
```

Comment:
Is this allowing missing columnTypes?  That should be an error. If the user doesn't want a column to be loaded, they should specify ColumnType.SKIP.